Reference implementation for torch.Tensor.sum_to_size

IvanYashchuk opened the pull request:
New ref: `torch._refs.sum_to_size`.

View consistency validation is disabled because the ref returns a view instead of returning the input.

IvanYashchuk committed:
Add error_inputs_sum_to_size

IvanYashchuk committed:
Add refs.sum_to_size

IvanYashchuk committed:
Add sum_to_size to torch_to_refs_map

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -1879,6 +1880,39 @@ def sum(
     )
 
 
+def _is_expandable_to(shape, desired):
```
Let's put this into utils?

mruberry replied:
Add type annotations

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -1879,6 +1880,39 @@ def sum(
 
+def _is_expandable_to(shape, desired):
+    # Python implementation of
+    # aten/src/ATen/ExpandUtils.h:is_expandable_to
```
Let's still comment here on what the function does

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -1879,6 +1880,39 @@ def sum(
+    if len(shape) > len(desired):
+        return False
+    for i in range(len(shape)):
+        if shape[i] != desired[i] and shape[i] != 1:
```
Doesn't this need to go from right-to-left?

IvanYashchuk replied:
Yes, it should be right-to-left, I wonder how tests pass 🤔

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -1879,6 +1880,39 @@ def sum(
+    )
+    # In ATen scalar tensors are sent through sum and the result is returned as
+    # type promoted
+    if shape == a.shape and len(shape) > 0:
```
I think there's a shape comparison util this should prefer using

IvanYashchuk replied:
`utils.is_same_shape` - very useful!

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -6146,6 +6146,14 @@ def sample_inputs_sum_to_size(op_info, device, dtype, requires_grad, **kwargs):
     return samples
 
+
+def error_inputs_sum_to_size(op_info, device, **kwargs):
```
Nice error input extension

IvanYashchuk committed:
Add _refs.sum_to_size to not_in_decomp_table

IvanYashchuk committed:
Move is_expandable_to to prims_common

IvanYashchuk committed:
Add one more error input

IvanYashchuk committed:
Check right-to-left

IvanYashchuk committed:
Use is_same_shape

IvanYashchuk committed:
Add more sample inputs; xfail conj/neg view tests

IvanYashchuk committed:
Remove f

IvanYashchuk committed:
flake8

IvanYashchuk committed:
Give up on making mypy happy with *shape

IvanYashchuk commented on torch/_refs/__init__.py:
```diff
@@ -1879,6 +1880,28 @@ def sum(
 
+def sum_to_size(
+    a: Tensor,
+    *shape,
```
What's the correct type annotation here?

mruberry replied:
We haven't been type annotating the `*shape` so you needn't bother

IvanYashchuk commented:
I fixed the problem with "is_expandable_to" and added error sample input to verify that the implementation is correct now.

@mruberry this is ready for another round.

mruberry reviewed (approved):
Awesome!

IvanYashchuk commented:
@pytorchbot merge -g
